### ISSUE TITLE
Order of clause in one-line if/else statement yields different mypy behaviors

### ISSUE THREAD

**Issue by fgmehlin** [python/mypy#8519]
I may have encountered a bug in Mypy's inference of types when using a one-line if/else statement:

The issue appears when returning a lambda function or a standard built-in (in that case `print`), depending on a `bool`.

```
import typing as t

def get_print_fn(quiet: bool) -> t.Callable[..., None]:
    """Return the standard print function if quiet is False, else an empty function."""
    return (lambda *args, **kwargs: None) if quiet else print
```

^^^ The above raises the following Mypy error:
`Incompatible return value type (got "function", expected "Callable[..., None]")`

However, reversing the condition like below does not show any error.

```
import typing as t

def get_print_fn(quiet: bool) -> t.Callable[..., None]:
    """Return the standard print function if quiet is False, else an empty function."""
    return print if not quiet else lambda *args, **kwargs: None
```

This somehow suggests that mypy is using the type of the `if` block to infer the return type? I'm not sure whether this is intended or indeed a bug.

**Meta information**

 - mypy version: `0.761`
 - `mypy.ini`

```
[mypy]
python_version = 3.7
warn_return_any = True
warn_unused_ignores = True
warn_unreachable = True
namespace_packages = True
show_column_numbers = True
strict_equality = True
no_implicit_optional = True
disallow_untyped_defs = True
disallow_any_generics = True
disallow_untyped_calls = True
```

**Comment by msullivan** [comment]
The issue here is that typechecking of conditional expressions works by computing the type of the if branch, and then using that as "context" when computing the type of the else branch. When print's type is used as context, that causes the lambda to get inferred as that type, but if the lambda's type is the context, that doesn't change anything.

Then, the two types get joined, and the join ends up as `function` because the sides aren't `similar`. So maybe the bug here is that `def (*Any, **Any)` should be compatible with everything?

**Comment by AlexWaygood** [comment]
Both examples cause mypy to [emit an error](https://mypy-play.net/?mypy=latest&python=3.10&flags=allow-redefinition&gist=91762dd69a9d84188053fd8c1abd01c4) on 0.941:

```python 
import typing as t

def get_print_fn(quiet: bool) -> t.Callable[..., None]:
    """Return the standard print function if quiet is False, else an empty function."""
    return (lambda *args, **kwargs: None) if quiet else print  # error: Incompatible return value type (got "function", expected "Callable[..., None]")

def get_print_fn2(quiet: bool) -> t.Callable[..., None]:
    """Return the standard print function if quiet is False, else an empty function."""
    return print if not quiet else lambda *args, **kwargs: None   # error: Incompatible return value type (got "function", expected "Callable[..., None]")
```

I think this *might* be due to changes to the `print` stub in typeshed.
